upstairs/src/downstairs.rs

```diff
@@ -0,0 +1,10875 @@
+        ));
+
+        // Process the error for client 1
+        assert!(!ds.process_ds_completion(
+            write_id,
+            ClientId::new(1),
+            err_response,
+            &UpstairsState::Active,
+            None,
+        ));
+
+        // process_ds_completion should return false, because the job was
+        // already acked (since it's a write)
+        assert!(!ds.process_ds_completion(
+            write_id,
+            ClientId::new(2),
+            ok_res,
+            &UpstairsState::Active,
+            None
+        ));
+
+        // Verify client states
+        assert_eq!(ds.clients[ClientId::new(0)].state(), DsState::Active);
+        assert_eq!(ds.clients[ClientId::new(1)].state(), DsState::Faulted);
+        assert_eq!(ds.clients[ClientId::new(2)].state(), DsState::Active);
+
+        // Verify we can ack this work
+        assert_eq!(ds.ackable_work().len(), 1);
+
+        // Verify the read switched from new to skipped
+
+        let job = ds.ds_active.get(&read_id).unwrap();
+
+        assert_eq!(job.state[ClientId::new(0)], IOState::InProgress);
+        assert_eq!(job.state[ClientId::new(1)], IOState::Skipped);
+        assert_eq!(job.state[ClientId::new(2)], IOState::InProgress);
+
+        assert_eq!(ds.clients[ClientId::new(0)].skipped_jobs.len(), 0);
+        assert_eq!(ds.clients[ClientId::new(1)].skipped_jobs.len(), 1);
+        assert_eq!(ds.clients[ClientId::new(2)].skipped_jobs.len(), 0);
+    }
+
+    #[tokio::test]
+    #[should_panic]
+    async fn deactivate_ds_not_when_initializing() {
+        // No deactivate of downstairs when upstairs not active.
+        let mut ds = Downstairs::test_default();
+
+        // This should panic, because `up` is in the wrong state
+        ds.try_deactivate(ClientId::new(0), &UpstairsState::Initializing);
+    }
+
+    #[tokio::test]
+    async fn write_fail_skips_many_jobs() {
+        // Create a bunch of jobs, do some, then encounter a write error.
+        // Make sure that older jobs are still okay, and failed job was
+        // skipped.
+        let mut ds = Downstairs::test_default();
+        set_all_active(&mut ds);
+        let write_one = ds.create_and_enqueue_generic_write_eob(false);
+        for i in ClientId::iter() {
+            ds.in_progress(write_one, i);
+        }
+
+        // Now, add a read.
+        let (read_one, request) = ds.create_and_enqueue_generic_read_eob();
+        for i in ClientId::iter() {
+            ds.in_progress(read_one, i);
+        }
+
+        // Make the read ok response
+        let rr = Ok(vec![ReadResponse::from_request_with_data(&request, &[])]);
+
+        for cid in ClientId::iter() {
+            ds.process_ds_completion(
+                write_one,
+                cid,
+                Ok(vec![]),
+                &UpstairsState::Active,
+                None,
+            );
+            ds.process_ds_completion(
+                read_one,
+                cid,
+                rr.
```

Comment:
I dont believe this condition exists any longer.
We go from LIveRepairReady -> LIveRepair and submit the first round of repair jobs at the same
step, so it's not really possible any longer to do this.

Reply:
Yeah agreed.